mathjax3-ts/output/chtml/Wrapper.ts

```diff
@@ -0,0 +1,604 @@
+    }
+
+    /*******************************************************************/
+
+    /*
+     * @override
+     */
+    constructor(factory: CHTMLWrapperFactory, node: MmlNode, parent: CHTMLWrapper = null) {
+        super(factory, node);
+        this.parent = parent;
+        this.bbox = BBox.zero();
+        this.getStyles();
+        this.getVariant();
+        this.getScale();
+        this.childNodes = node.childNodes.map((child: Node) => {
+            return this.wrap(child as MmlNode);
+        });
+    }
+
+    /*
+     * @param{MmlNode} node  The node to the wrapped
+     * @param{CHTMLWrapper} parent  The parent wrapped node
+     * @return{CHTMLWrapper}  The newly wrapped node
+     */
+    public wrap(node: MmlNode, parent: CHTMLWrapper = null) {
+        const wrapped = this.factory.wrap(node, parent || this);
+        if (parent) {
+            parent.childNodes.push(wrapped);
+        }
+        return wrapped;
+    }
+
+    /*******************************************************************/
+    /*
+     * Create the HTML for the wrapped node.
+     *
+     * @param{HTMLElement} parent  The HTML node where the output is added
+     * @param{number[]} WHD  Either [W] or [H,D], giving the dimensions to stretch
+     *                       a stretchable embellished operator to
+     */
+    public toCHTML(parent: HTMLElement, WHD: number[] = []) {
+        let chtml = parent;
+        if (!this.node.isInferred) {
+            chtml = this.standardCHTMLnode(parent);
+        }
+        for (const child of this.childNodes) {
+            child.toCHTML(chtml, WHD);
+        }
+    }
+
+    /*******************************************************************/
+    /*
+     * Return the wrapped node's bounding box, either the cached one, if it exists,
+     *   or computed direcctly if not.
+     *
+     * @param{boolean} save  Whether to cache the bbox or not (used for stretchy elements)
+     * @return{BBox}  The computed bounding box
+     */
+    public getBBox(save: boolean = true) {
+        if (!this.bboxComputed) {
+            let bbox = this.computeBBox();
+            if (save) {
+                this.bbox = bbox;
+                this.bboxComputed = true;
+            }
+            return bbox;
+        }
+        return this.bbox;
+    }
+
+    /*
+     * @return{BBox}  The computed bounding box for the wrapped node
+     */
+    protected computeBBox() {
+        const bbox = BBox.empty();
+        for (const child of this.childNodes) {
+            bbox.append(child.getBBox());
+        }
+        bbox.clean();
+        return bbox;
+    }
+
+    /*******************************************************************/
+
+    /*
+     * Add the style attribute, but remove any font-related styles
+     *   (since these are handled separately by the variant)
+     */
+    protected getStyles() {
+        const styleString = this.node.attributes.getExplicit('style') as string;
+        if (styleString) {
```

Comment:
rewrite to 
if (!styleString) {
return;
}
to save one layer of indentation.

Reply:
But this introduces an extra operation (the negation), which impacts performance (slightly).  Since `getStyles()` is used for every element being displayed, I would rather have the indentation than the extra operation.

Reply:
I also prefer there to be only one exit to a function, when possible, rather than multiple ones.

Reply:
Testing for bit existence or non-existence should not really impact performance.
I would rather keep nesting down to improve code readability. It's a "philosophical" thing.

In general, my philosophy is to concentrate on the main logic in a code block and dispense with the special cases beforehand (return immediately or continue in a loop). That generally keeps nesting down and it often makes it easier to see that one has covered all special cases.

Reply:
OK, I'll do it if I can use

    if (!styleString) return;

The extra bracing and line breaks make it harder for me to read, so if this is about readability, I would like the one-line version.

Reply:
Sounds good to me.